### DISCUSSION TITLE
rehype-sanitize has different output than github

### DISCUSSION THREAD

**Discussion by RARgames** [rehypejs/rehype#159] [Q&A] [answered]
From readme:
> It drops anything that isn’t explicitly allowed by a schema (defaulting to how github.com works)

Example:
input:
```
<!-- test --> 123

<!-- test --> 456 <!-- test -->
```
output github:
`123 456 `
output rehype-sanitize:
empty

**Comment by ChristianMurphy** [comment, marked as answer]
Hey @RARgames! 👋 
Sorry you ran into some confusion.

rehype-santize does behave as GitHub does.
Make sure you have your remark and rehype configuration setup correctly.
A working example for reference: https://stackblitz.com/edit/github-ozn1qm?file=src%2Fmain.ts

<details>

<summary>source from sandbox</summary>

```ts
// ⚠️ Important! Please make sure the dependencies are up to date.

import { unified } from 'unified';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import rehypeRaw from 'rehype-raw';
import rehypeSanitize from 'rehype-sanitize';
import rehypeStringify from 'rehype-stringify';

const sourceMarkdown = `
<!-- test --> 123

<!-- test --> 456 <!-- test -->
`;

try {
  document.querySelector<HTMLPreElement>('#source')!.textContent =
    sourceMarkdown;

  const file = await unified()
    .use(remarkParse)
    // Add any remark plugins here
    .use(remarkRehype, { allowDangerousHtml: true })
    // Add any rehype plugins here
    .use(rehypeRaw)
    .use(rehypeSanitize)
    .use(rehypeStringify)
    .process(sourceMarkdown);

  document.querySelector<HTMLIFrameElement>(
    '#result'
  )!.contentWindow!.document.body.innerHTML = String(file);
  document.querySelector<HTMLPreElement>('#error')!.textContent = '';
} catch (error) {
  document.querySelector<HTMLPreElement>('#error')!.textContent = String(error);
}

```

</details>

**Reply by RARgames** [reply]
Thanks for the fast reply. It's probably some deeper issue in my code or with the editor I'm using/customizing: react-md-editor.

**Reply by ChristianMurphy** [reply]
A few considerations to start with:
`@uiw/react-md-editor` builds on  https://github.com/uiwjs/react-markdown-preview
* `react-markdown-preview` is a major version behind in all unified/remark/rehype related dependencies
* `react-markdown-preview` adds https://github.com/jaywcjlove/rehype-attr which also uses/changes comments
* `react-markdown-preview` also adds https://github.com/jaywcjlove/rehype-ignore which uses/changes comments

**Reply by RARgames** [reply]
Thanks for ideas!

Unfortunately, I couldn't find what is causing the issue yet.

I created a custom version of `react-markdown-preview` and removed remark, rehype plugins. Also I updated markdown-related dependencies. Final dependencies:
```
"dependencies": {
  "@babel/runtime": "^7.17.2",
  "@uiw/copy-to-clipboard": "~1.0.12",
  "react-markdown": "~9.0.1",
  "unist-util-visit": "^5.0.0"
}
```
Also I tested it on the demo version of `react-md-editor` so I know that issue is not in my code.

The issue still exists and as far as I know is not related to:
- my code
- `rehype-attr`
- `rehype-ignore`
- old unified/remark/rehype related dependencies

Please let me know if you have any other ideas.

In the meantime, I'll try to figure out what might causing the issue myself.

**Reply by ChristianMurphy** [reply]
It has to do with customization made to the editor. Keep digging.
They also set `skipHTML` and `allowElement` properties on `react-markdown` https://github.com/uiwjs/react-markdown-preview/blob/74fa8e98070a6ceaeef8055975aefaa3a43e3d4e/core/src/preview.tsx#L45-L67
which could be related, they also change other properties.

They also have their own issue tracker where they may be able to point you in the right direction https://github.com/uiwjs/react-markdown-preview/issues

**Reply by RARgames** [reply]
Thanks! Will do.
I'm already in contact with the author on `react-md-editor` issue tracker - pointed me to rehype/remark plugins as well.

**Reply by RARgames** [reply]
It seems that the issue might be connected to `rehype-sanitize` or/and `react-markdown`. This is an example from the author of `react-md-editor`: https://codesandbox.io/embed/sharp-mountain-m7yx7t

It's similar to the way that plugins are used in the source code of `react-md-editor`, `react-markdown-preview`.

I know that your example (https://stackblitz.com/edit/github-ozn1qm?file=src%2Fmain.ts) works.

**Reply by RARgames** [reply]
This is a working example:

```js
import Markdown from "react-markdown";
import rehypeSanitize from "rehype-sanitize";
import rehypeRaw from "rehype-raw";

const source = `
## MarkdownPreview

**Hello world!!!** <IFRAME SRC=\"javascript:javascript:alert(window.origin);\"></IFRAME>

<!-- test --> 123

<!-- test --> 456 <!-- test -->
`;

const rehypePlugins = [rehypeRaw, rehypeSanitize];

export default function App() {
  return (
    <div className="App">
      <Markdown rehypePlugins={rehypePlugins}>
        {source}
      </Markdown>
    </div>
  );
}
```
Why do we need rehypeRaw before rehypeSanitize?
Do other plugins you listed e.g. `remarkParse`, `remarkRehype` help in any way? (Prevent some random behavior like this issue)

**Reply by wooorm** [reply]
> Why do we need rehypeRaw before rehypeSanitize?

Because plugins are order sensitive. Like so many things in life, doing two arbitrary things, doing them in a different order doesn’t work.

**Reply by RARgames** [reply]
My bad, the question was not precise: Why do we need to use rehypeRaw at all in order to make rehypeSanitize work correctly?
Proposed answer: We need to re-parse the document in order to rehypeSanitize has access to particular nodes instead of merged nodes (like it happens in case of comments, when there is 1 node instead of 3 here: `<!-- test --> 456 <!-- test -->`).

Is this right?

**Reply by wooorm** [reply]
Yes

**Reply by RARgames** [reply]
Sorry to ask similar questions over and over again. :)

I'm finishing my pipeline, but still cannot figure out how to make it work correctly.

I have 2 conflicting plugins: `rehype-raw` and `rehype-prism-plus`.
Either I have fully-working code blocks (comments are not working as intended), or I have fully-working comments (code blocks are not working as intended).

I have a few other rehype/remark plugins and a custom schema for rehype-sanitize, but they are working perfectly fine.

Here is the minimal example (without other plugins etc.)
https://codesandbox.io/embed/vq963l?view=Editor+%2B+Preview&module=%2Fsrc%2FApp.js

![image](https://github.com/rehypejs/rehype/assets/13639766/c53c2529-6a37-47ca-b158-91123fee4170)

Example 1 is almost good, but it's missing classes: `line-number`, `highlight-line` and `line="number"`. Probably `rehype-raw` is removing `{1,3-4} showLineNumbers` from the start of the code block.
AFAIK `rehype-prism-plus` is getting this info ({1,3-4} showLineNumbers) from meta data: `node.data.meta` or `node.properties.metastring`.

Probably I can somehow save this info (e.g. using other plugin to create a node before `rehype-raw`) and then retrieve the info after `rehype-raw` using another plugin - but it seems like an overkill.

Any help will be appreciated!

**Reply by RARgames** [reply]
I've managed to get it working:

- I created 2 plugins: `preserveMetaPlugin` and `retrieveMetaPlugin`
- This is my minimal example plugins pipeline: `[preserveMetaPlugin, rehypeRaw, metaBackPlugin, rehypeSanitize, rehypePrismPlus]`

Not sure if it is the best way, but it works.

```js
function preserveMetaPlugin() {
    return (tree, file) => {
      visit(tree, 'element', (node) => {
        if (node.tagName === 'code' && node.data && node.data.meta) {
          node.properties = { ...node.properties, 'data-meta': String(node.data.meta) };
        }
      });
    };
  }

  function retrieveMetaPlugin() {
    return (tree, file) => {
      visit(tree, 'element', (node) => {
        if (node.tagName === 'code' && node.properties && node.properties['dataMeta']) {
            if (!node.data) {
              node.data = {};
            }
            node.data.meta = node.properties['dataMeta'];
            delete node.properties['dataMeta'];
        }
      });
    };
  }
```
